radio/src/targets/nv14/touch_driver.cpp

```diff
@@ -362,37 +375,134 @@ static void ft6x06_TS_GetXY(uint16_t DeviceAddr, uint16_t * X, uint16_t * Y, uin
+ *         Manage multi touch thanks to touch Index global
+ *         variable 'tc_handle.currActiveTouchIdx'.
+ * @param  DeviceAddr: Device address on communication Bus.
+ * @param  X: Pointer to X position value
+ * @param  Y: Pointer to Y position value
+ * @retval None.
+ */
+static void cst836u_TS_GetXY(uint16_t * X, uint16_t * Y, uint32_t * event)
+{
+  uint8_t regAddress = 0;
+  uint8_t dataxy[4];
+
+  if (tc_handle.currActiveTouchIdx < tc_handle.currActiveTouchNb) {
+    switch (tc_handle.currActiveTouchIdx) {
+      case 0 :
+        regAddress = CST836U_TOUCH1_XH_REG;
+        break;
+      case 1 :
+        regAddress = CST836U_TOUCH2_XH_REG;
+        break;
+      default :
+        break;
+    }
+
+    /* Read X and Y positions */
+    TS_IO_ReadMultiple(TOUCH_CST836U_I2C_ADDRESS, regAddress, dataxy, sizeof(dataxy));
+    /* Send back ready X position to caller */
+    *X = ((dataxy[0] & CST836U_MSB_MASK) << 8) | dataxy[1];
+    /* Send back ready Y position to caller */
+    *Y = ((dataxy[2] & CST836U_MSB_MASK) << 8) | dataxy[3];
+
+    *event = (dataxy[0] & CST836U_TOUCH_EVT_FLAG_MASK) >> CST836U_TOUCH_EVT_FLAG_SHIFT;
+    /*
+    uint32_t weight;
+    uint32_t area;
+    */
+    tc_handle.currActiveTouchIdx++;
+  }
+}
+
+/**
+ * @brief  Return if there is touches detected or not.
+ *         Try to detect new touches and forget the old ones (reset internal global
+ *         variables).
+ * @param  DeviceAddr: Device address on communication Bus.
+ * @retval : Number of active touches detected (can be 0, 1 or 2).
+ */
+static uint8_t cst836u_TS_DetectTouch()
+{
+  volatile uint8_t nbTouch = 0;
+
+  /* Read register CST836U_TOUCH_NUM_REG to check number of touches detection */
+  nbTouch = TS_IO_Read(CST836U_TOUCH_NUM_REG);
+  if (nbTouch > CST836U_MAX_DETECTABLE_TOUCH) {
+    /* If invalid number of touch detected, set it to zero */
+    nbTouch = 0;
+  }
+  tc_handle.currActiveTouchNb = nbTouch;
+
+  tc_handle.currActiveTouchIdx = 0;
+  return (nbTouch);
+}
+
 void TouchReset()
 {
   LL_GPIO_ResetOutputPin(TOUCH_RST_GPIO, TOUCH_RST_GPIO_PIN);
-  delay_ms(20);
+  delay_ms(10);
   LL_GPIO_SetOutputPin(TOUCH_RST_GPIO, TOUCH_RST_GPIO_PIN);
   delay_ms(300);
 }
 
+void detectTouchController()
+{
+  touchController = TC_FT6236;
+  if( stm32_i2c_is_dev_ready(TOUCH_I2C_BUS, TOUCH_CST836U_I2C_ADDRESS, 3, 5) == 0)
+      touchController = TC_CST836U;
+}
+
 void TouchInit(void)
 {
   I2C_Init();
   TouchReset();
-  touch_ft6236_debug_info();
-  /* INT generation for new touch available */
-  /* Note TS_INT is active low */
-  uint8_t regValue = 0;
-  regValue = (FT6206_G_MODE_INTERRUPT_TRIGGER & (FT6206_G_MODE_INTERRUPT_MASK >> FT6206_G_MODE_INTERRUPT_SHIFT)) << FT6206_G_MODE_INTERRUPT_SHIFT;
-  /* Set interrupt TOUCH_FT6236_I2C_ADDRESS mode in FT6206_GMODE_REG */
-  TS_IO_Write(TOUCH_FT6236_I2C_ADDRESS, FT6206_GMODE_REG, regValue);
+  detectTouchController();
+  switch(touchController)
```

Comment:
Ideally the detection function would set a group of function pointer to be used a generic way by the driver instead of testing this all the time.

Reply:
I will postpone this to a complete touch driver restructuring